### ISSUE TITLE
Proposal: Add `not`

### ISSUE THREAD

**Issue by younho9** [sindresorhus/ts-extras#32]
```ts
const isNullable = (value: unknown): value is null | undefined =>
	value === null || value === undefined;

const isNonNullable = not(isNullable);

declare const someValue: string | null | undefined;

if (isNonNullable(someValue)) {
	someValue;
	//=> string
}
```

It could be useful to reuse existing predicate.

use-case: [TypeScript Playground](https://www.typescriptlang.org/play?jsx=0#code/C4TwDgpgBACgThAJgSwMYENgQDwBVwQB8UAvFABQBu6ANgK4QBcUdAdgNasD2A7qwJSli1etGQBnKPkgBuALAAoRQDM2qYMi6so3YHgKFyYBCgxZm8JGkw5pRQQG9FUF1ATA6cbdgBqtBoYiDMx+ovzMQWKSAKIAHqj0iDihDAA0UgZCUACExlZmEFT+EPzyCgC+ZYqoWuLAUBIAcnQ0NOgARjTQZEWizGycvAIRxQ2SrC00UAA+LKxJysisSEKKAJCRpCRkE60zs5vbZGwLS0hVCjWsdTqTHV0AgnBw6CCkUADauzTpJxCLy0Q6QA5AALCCtLjAgC6FyuN3QzwAjO9vvcIE8XiAAHSLGhYODkJp3TolGRQAD0FJIxHI3xmc1OgP4H2h1Vq9URcAATKiSY9nq9cch8RBCboieJmq10fxSpTqcQ6nAlgBzVlAA)

**Comment by jonahsnider** [comment]
Is inverting the return type of a type guard possible with TypeScript's type system?

**Comment by younho9** [comment]
It depends on `Exclude` type. Type guard using `Exclude` can exclude some types from a given type.

https://github.com/microsoft/TypeScript/issues/4183#issuecomment-364688498

```ts
type Predicate<Type> = (value: unknown) => value is Type;

function not<Type>(predicate: Predicate<Type>) {
    return <Value>(value: Value): value is Exclude<Value, Type> => !predicate(value);
};

const isNullable = (value: unknown): value is null | undefined =>
	value === null || value === undefined;

const isNonNulluable = not(isNullable);
//=> <Value>(value: Value) => value is Exclude<Value, null | undefined>
```

use-cases: 
- `Nullable` <-> `NonNullable`
- `Falsy` <-> `Truthy`

**Comment by younho9** [comment]
> @younho9 I don't think this will work with `isEmtpy` or `isInfinite`. It's not only depends on `Exclude` but also depends on negated type. [microsoft/TypeScript#29317](https://github.com/microsoft/TypeScript/pull/29317)

Yes, when this PR merge, the negated type guards will be more useful.

Negative type guards (with `Exclude`) do not narrow types in some situations.
- https://github.com/microsoft/TypeScript/issues/44901
- https://github.com/microsoft/TypeScript/issues/43589

But I still think some types guards with `Exclude` are useful.

**Comment by fregante** [comment]
Related type-only request:

- https://github.com/sindresorhus/type-fest/issues/417

**Comment by sindresorhus** [comment]
Good idea! We've added a `not()` helper that inverts type guards via `Exclude`:

```typescript
export function not<T>(predicate: (value: unknown) => value is T) {
    return <V>(value: V): value is Exclude<V, T> => !predicate(value);
}
```

Works great for common patterns like:
- Filtering arrays: `values.filter(not(isNullish))`
- Type narrowing: `if (not(isEmpty)(array)) { /* array is non-empty */ }`
- Composing predicates without duplicating logic

Added comprehensive tests and documentation noting the TypeScript limitations you mentioned (issues #44901 and #43589).

Implemented in commit 99d5ef1.
